File: get_comments.py
import praw
from google.cloud import texttospeech
import pydub
import os
from pydub import AudioSegment

# export GOOGLE_APPLICATION_CREDENTIALS=/Users/othmane/Development/tiktok-reddit-bot/tiktok-reddit-bot-key.json


reddit = praw.Reddit(
    client_id='Zl5cZsjPgTr1GW1U5T64sw',
    client_secret='mnNDGr2voE4jhGQeczLew4hTH1YVLQ',
    user_agent='macos:tiktok-reddit-bot:v1.0 (by /u/Lower-Present2622)'
)

def get_reddit_comments(post):

    VIDEO_TIME_LIMIT = 55
    submission = reddit.submission(url=post)

    # Replace the "MoreComments" objects with actual comments up to a certain level
    submission.comments.replace_more(limit=0)
    list_comments = [submission.title]
    full_text = submission.title

    for i, top_level_comment in enumerate(submission.comments):
        comment = top_level_comment.body
        temp_full_text = full_text + comment
        if (full_text + comment).count(" ") > 131 : # number of spaces + 1
            break
        else :
            full_text += " " + comment
            list_comments.append(comment)
    
    audio_file = "full_speech.mp3"
    full_speech = comment_to_speach(full_text, audio_file)

    with open(f"audio_files/full_speech.mp3", "wb") as out:
        out.write(full_speech.audio_content)

    audio_length = pydub.AudioSegment.from_file(f"audio_files/full_speech.mp3", format="mp3").duration_seconds

    # split sentences every three word
    final_list_comments = [cut_sentence(comment) for comment in list_comments]


    return {
        "url": post,
        "length" : audio_length,
        "audio_file": audio_file,
        "comments" : final_list_comments
    }







def comment_to_speach(text_to_convert, output_name):

    # Instantiates a client
    client = texttospeech.TextToSpeechClient()

    # Set the text input to be synthesized
    synthesis_input = texttospeech.SynthesisInput(text=text_to_convert)

    # Build the voice request, select the language code and the ssml voice gender
    voice = texttospeech.VoiceSelectionParams(
        language_code="en-US", ssml_gender=texttospeech.SsmlVoiceGender.MALE
    )

    # Select the type of audio file you want returned
    audio_config = texttospeech.AudioConfig(
        audio_encoding=texttospeech.AudioEncoding.MP3
    )

    # Perform the text-to-speech request on the text input with the selected
    # voice parameters and audio file type
    response = client.synthesize_speech(
        input=synthesis_input, voice=voice, audio_config=audio_config
    )

    response

    return response

    """# The response's audio_content is binary.
    with open(f"audio_files/{output_name}", "wb") as out:
        out.write(response.audio_content)"""




def cut_sentence(sentence):
    words = sentence.split(" ")
    cut_sentences = []
    for i in range(0, len(words), 3):
        cut_sentence = " ".join(words[i : i + 3])
        cut_sentences.append(cut_sentence)
    return cut_sentences

#print(get_reddit_comments('https://www.reddit.com/r/explainlikeimfive/comments/14wytj0/eli5_how_does_nasa_ensure_that_astronauts_going/'))

File: get_comments.go
package main


import (
	"context"
	"fmt"
	"net/url"
	"log"
	"io/ioutil"
	"strings"
	"github.com/vartanbeno/go-reddit/v2/reddit"
	texttospeech "cloud.google.com/go/texttospeech/apiv1"
        "cloud.google.com/go/texttospeech/apiv1/texttospeechpb"
)



func RetrieveSubredditAndPostId(link string) (string, string ,error) {
	parsedURL, err := url.Parse(link)
	if err != nil {
		return "", "", err
	}

	pathParts := strings.Split(parsedURL.Path, "/")
	if len(pathParts) < 5 {
		return "", "", fmt.Errorf("Invalid reddit post link")
	}
	/*  /r/{subreddit_name}/comments/{post_id}/{slug}/
		[0] "" | [1] "r" | [2] "{subreddit_name}" | [3] "comments"
		[4] "{post_id}" | [5] "{slug}" | [6] ""
	*/
	return pathParts[2], pathParts[4], nil
}

func GetComments(subreddit_name string, id string) (string, string, []string, error){

	client, err := reddit.NewReadonlyClient()
	if err != nil {
		return "", "", nil, err
	}

	comments, _, err := client.Post.Get(context.Background(), id)
	var commentBodies []string
	commentBodies = []string{
		comments.Post.Title,
	}

	var full_text string 
	full_text = comments.Post.Title
	for _, comment := range comments.Comments {
		if !(strings.Count((full_text + comment.Body), " ") > 160) {
			commentBodie2Add := comment.Body
			commentBodies = append(commentBodies, commentBodie2Add)
			full_text = full_text + comment.Body
		} else {
			break
		}
	}

	// synthesize the speech and write it to audio file
	speech := SynthesizeSpeech(full_text)

	// The resp's AudioContent is binary.
	filename := "speech.mp3"
	var file_path string = fmt.Sprintf("audio_files/%s", filename)
	outcome := ioutil.WriteFile(file_path, speech.AudioContent, 0644)

	if outcome != nil {
			log.Fatal(outcome)
	}
	fmt.Printf("Audio content written to file: %v\n", file_path)

	return filename, full_text, commentBodies, nil

}

func SynthesizeSpeech(text string) (*texttospeechpb.SynthesizeSpeechResponse){
	// Instantiates a client.
	ctx := context.Background()

	client, err := texttospeech.NewClient(ctx)
	if err != nil {
			log.Fatal(err)
	}
	defer client.Close() // defer means that the action will be done right before return

	// Perform the text-to-speech request on the text input with the selected
	// voice parameters and audio file type.
	req := texttospeechpb.SynthesizeSpeechRequest{
			// Set the text input to be synthesized.
			Input: &texttospeechpb.SynthesisInput{
					InputSource: &texttospeechpb.SynthesisInput_Text{Text: text},
			},
			Voice: &texttospeechpb.VoiceSelectionParams{
					LanguageCode: "en-US",
					SsmlGender:   texttospeechpb.SsmlVoiceGender_MALE,
			},
			// Select the type of audio file you want returned.
			AudioConfig: &texttospeechpb.AudioConfig{
					AudioEncoding: texttospeechpb.AudioEncoding_MP3,
			},
	}

	response, err := client.SynthesizeSpeech(ctx, &req)
	if err != nil {
			log.Fatal(err)
	}

	return response 


}

func splitEveryNWords(s []string, n int) ([]string) {
	var chunks []string
	
	for _, sentence := range s {
		words := strings.Split(sentence, " ")
		fmt.Print("this is words  ")
		fmt.Println(words)
		var compteur int = 0  
		for _,word := range words {
			fmt.Println(word)
			if compteur+3 >= len(words){
				joined := strings.Join(words[compteur:], " ")
				//fmt.Println(joined)
				chunks = append(chunks, joined)
				break
			} else {
				joined := strings.Join(words[compteur:compteur+3], " ")
				chunks = append(chunks, joined)
				compteur = compteur + 3
			}
		}
	}
	

	return chunks
}


func main() {
	subreddit, postID, err_retrieve := RetrieveSubredditAndPostId("https://www.reddit.com/r/explainlikeimfive/comments/14wytj0/eli5_how_does_nasa_ensure_that_astronauts_going/")
	if err_retrieve != nil {
		log.Fatal(err_retrieve)
	}
	speeche, _, comments, err := GetComments(subreddit, postID)
	if err != nil {
		log.Fatal(err)
	}

	fmt.Println(comments)
	fmt.Println("==========================")

	parsed_comments := splitEveryNWords(comments, 3)

	subtitles_path, err := createSubtitlesFile("subtitles.srt", parsed_comments)
	fmt.Println(subtitles_path)
	//fmt.Println("\n \n" + text)
	//fmt.Println(speech)

	video_extract_path, _ := CutVideoExtract("minecraft_1.mp4")
	fmt.Println(video_extract_path)
	fmt.Println(err)
	video_audio_extract_path, err := MergeVideoAudio(video_extract_path, speeche)
	fmt.Println(video_audio_extract_path)
	fmt.Println(err)
	final_tiktok_video, err := BurnSubtitles(video_audio_extract_path, "subtitles.srt")
	fmt.Println(final_tiktok_video)
	fmt.Println(err)
}